### PR TITLE
[SPARK-22973][SQL] Fix incorrect results of Casting Map to String

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr fixed the issue when casting maps into strings;
```
scala> Seq(Map(1 -> "a", 2 -> "b")).toDF("a").write.saveAsTable("t")
scala> sql("SELECT cast(a as String) FROM t").show(false)
+----------------------------------------------------------------+
|a                                                               |
+----------------------------------------------------------------+
|org.apache.spark.sql.catalyst.expressions.UnsafeMapData@38bdd75d|
+----------------------------------------------------------------+
```
This pr modified the result into;
```
+----------------+
|a               |
+----------------+
|[1 -> a, 2 -> b]|
+----------------+
```

## How was this patch tested?
Added tests in `CastSuite`.

### CODE REVIEW

**Comment by maropu** [comment]
@cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 1]
```diff
@@ -228,6 +228,35 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
...
         builder.build()
       })
+    case MapType(kt, vt, _) =>
+      buildCast[MapData](_, map => {
+        val builder = new UTF8StringBuilder
+        builder.append("[")
+        if (map.numElements > 0) {
+          val keyToUTF8String = castToString(kt)
+          val valueToUTF8String = castToString(vt)
+          builder.append(keyToUTF8String(map.keyArray().get(0, kt)).asInstanceOf[UTF8String])
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: `map.keyArray()` and `map.valueArray` appear many times, we can create 2 local variables for them.

**Comment by cloud-fan** [comment]
LGTM except one minor comment

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 1 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: ok

**Comment by maropu** [comment]
BTW, the current `Datset.showString` prints rows thru `RowEncoder` deserializers like;
```
scala> Seq(Seq(Seq(1, 2), Seq(3), Seq(4, 5, 6))).toDF("a").show(false)
+------------------------------------------------------------+
|a                                                           |
+------------------------------------------------------------+
|[WrappedArray(1, 2), WrappedArray(3), WrappedArray(4, 5, 6)]|
+------------------------------------------------------------+
```
If [we  cast them before prints](https://github.com/apache/spark/compare/master...maropu:CastToStringInShowString), we could get more simpler forms like;
```
scala> Seq(Seq(Seq(1, 2), Seq(3), Seq(4, 5, 6))).toDF("a").show(false)
+------------------------+
|a                       |
+------------------------+
|[[1, 2], [3], [4, 5, 6]]|
+------------------------+
```
I'm not sure through, is this acceptable? (Probably, we might need to add a option to keep the old behaviour)

**Comment by cloud-fan** [comment]
I think this is a bug, in `Dataset.showString` I see code like
```
case seq: Seq[_] => seq.mkString("[", ", ", "]")
```
Which means we do want to show strings like `[[1, 2], [3], [4, 5, 6]]`

Anyway let's fix in another PR, I'm merging this PR first

**Comment by cloud-fan** [comment]
merging to master/2.3!

**Comment by maropu** [comment]
ok, I'll fix struct in a next following pr first.

**Pull request closed without merging**